### ISSUE TITLE
Embedded programming languages

### ISSUE THREAD

**Issue by E5war5IT** [microsoft/language-server-protocol#1757]
I need a very interesting answer that torments me, I will say that the documentation is not so well done. It looks like you did it for yourself.

How do I make an embedded programming language, for example, in Rust, C++? That is, something is needed to make an embedded programming language work after the main programming language if it is an interpreter or compiler.

**Comment by dbaeumer** [comment]
The language server protocol doesn't provide any means right now for embedded languages. We do provided a special guide how to achieve this if you write a language server for VS Code (https://code.visualstudio.com/api/language-extensions/embedded-languages). The main reason is that language embedding depends on how clients manage language support.

Hope that helps.

**Comment by E5war5IT** [comment]
This problem is not solved, because since it does not describe to me how to do. You just described the details in two words.

For example, I still don't understand how to make my LSP (https://code.visualstudio.com/api/language-extensions/language-server-extension-guide)

For example, how to make variables that the developer himself created and functions, constants and much more displayed. How to display hints after the instructions, for example, what you can write, make documentation for the instructions. 

I've been dealing with this issue for a very long time that it's not even a month.

A lot of questions and zero answers.

**Comment by dbaeumer** [comment]
Examples both for LSP servers and embedded languages can be found here: https://github.com/microsoft/vscode-extension-samples

**Comment by angelozerr** [comment]
@dbaeumer I think really that LSP should provide the capability to support embedded language. The HTML language server sample works because it embedds CSS, JavaScript language service inside the language server. This strategy have some limitation:

 * when your language server is written with another language than TypeScript, you cannot manage this strategy. For instance vscode-xml uses LemMinx, an XML Language server written in Java, we cannot use CSS, JavaScript language service inside the XML Language service.
 * if you want to manage in HTML Language server another language with `<script language="java">` it will very hard to implement this feature since Java Language server is written in Java.

In other words I think LSP should provide the capability to delegate to an another language server the LSP operations. The idea is that the language server should provide an embedded range with a given language. The LSP client could consumes `textDocument/embeddedRanges` which return a list of `{Range, embeddedLangeId}` and after that LSP client delegate the LSP operation according the embedded range/languageId.

**Comment by dbaeumer** [comment]
Unfortunately this is not so easy since the embedded part can usually not be send to the another language server 1:1 since they are almost always fragments which might not be syntactically correct. So some rewriting and position mapping needs to be done and that needs to be implemented by the server. 

What we did discuss is to offer a way that a server A can request a LSP operation in server B via the client to make this from an implementation point of view easier. But in principal this is already possible via custom messages.
